services/QuillLMS/app/workers/assign_recommendations_worker.rb

```diff
@@ -4,53 +4,68 @@ class AssignRecommendationsWorker
   include Sidekiq::Worker
   sidekiq_options queue: SidekiqQueue::CRITICAL
 
-  # rubocop:disable Metrics/CyclomaticComplexity
-  def perform(options={})
-    options = options.with_indifferent_access
-    unit_template_id = options["unit_template_id"]
-    classroom_id = options["classroom_id"]
-    student_ids = options["student_ids"]
-    last = options["last"]
-    lesson = options["lesson"]
-    assign_on_join = options["assign_on_join"] || false
-    assigning_all_recommended_packs = options["assigning_all_recommended_packs"]
+  # rubocop:disable Metrics/ParameterLists
+  def perform(
+    classroom_id:,
+    lesson:,
+    student_ids:,
+    unit_template_id:,
+    activity_pack_sequence_id: nil,
+    assign_on_join: false,
+    assigning_all_recommendations: false,
+    is_last_recommendation: true,
+    order: nil
+  )
 
     classroom = Classroom.find(classroom_id)
     teacher = classroom.owner
     units = find_units(unit_template_id, teacher.id)
+    unit = nil
+
     if units.present?
       unit = find_unit(units)
-      unit.update(visible:true) if unit && !unit.visible
+      unit.update(visible: true) if unit && !unit.visible
     end
+
     classroom_data = {
       id: classroom_id,
       student_ids: student_ids,
       assign_on_join: assign_on_join
     }
-    unit ||= nil
+
     assign_unit_to_one_class(unit, classroom_id, classroom_data, unit_template_id, teacher.id)
+
+    unit = find_unit(find_units(unit_template_id, teacher.id)) if unit.nil?
```

Comment:
Is there a reason to do this here instead of line 23?  Looks like this is identical to the value set on 26, so we might as well just set this exact value earlier in the workflow rather than having the possibility of setting it in the `if` block on line 21 or set here to the same value if it's not set in the `if` block.

Reply:
Before explaining this logic, I'd like to say I didn't fully refactor this code as much I wanted, so there's more refactoring opportunity here, specifically the clunky logic flow of `find_unit` and find_units` that you are inquiring about.

Let's say were at line 38.  There are two cases

Case 1:  `unit` is not nil.  
Then line 36 doesn't matter doesn't get called and we move along.

Case 2:  unit is `nil` 
Then Units::Creator will be called on line 55 from the `assign_unit_to_one_class` on line 36.  This creates a new unit that will be assigned in line 36.  We can't make this assignment earlier since the unit doesn't exist yet and we need to pass it to `save_activity_pack_sequence_activity_pack` immediately following.